composer/datasets/cifar.py

```diff
@@ -80,6 +86,106 @@ def build_cifar10_dataloader(
+    batch_size: int,
+    is_train: bool = True,
+    download: bool = True,
+    drop_last: bool = True,
+    prefetch_factor: int = 2,
+    num_workers: int = 8,
+    ffcv_dir: str = '/tmp',
+    ffcv_dest: str = 'cifar_train.ffcv',
+    ffcv_write_dataset: Union[str, bool] = False,
+    datadir: Union[str, None] = None,
+):
+    """Builds and FFCV CIFAR10 dataloader
+
+    Args:
+        batch_size (int): Batch size per device
+        is_train (bool): Whether to load the training data or validation data. Default:
+            ``True``.
+        download (bool, optional): Whether to download the dataset, if needed. Default:
+            ``True``.
+        drop_last (bool): Drop remainder samples. Default: ``True``.
+        drop_last (bool): whether to drop last samples. Default: ``True``.
+        prefetch_factor (int): Number of batches to prefect. Default: ``2``.
+        ffcv_dir (str, optional): A directory containing train/val <file>.ffcv files. If these files don't exist and
+            ``ffcv_write_dataset`` is ``True``, train/val <file>.ffcv files will be created in this dir. Default: ``"/tmp"``.
+        ffcv_dest (str, optional): <file>.ffcv file that has dataset samples. Default: ``"cifar_train.ffcv"``.
+        ffcv_write_dataset (str | bool, optional): Whether to create dataset in FFCV format (<file>.ffcv) if it doesn't exist. Default:
+        ``False``.
+        datadir (str | None, optional): Path to the non-FFCV data directory
+    """
+    try:
+        import ffcv
+        from ffcv.fields.decoders import IntDecoder, SimpleRGBImageDecoder
+        from ffcv.pipeline.operation import Operation
+    except ImportError:
+        raise ImportError(
+            textwrap.dedent("""\
+            Composer was installed without ffcv support.
+            To use ffcv with Composer, please install ffcv in your environment."""))
+
+    dataset_filepath = os.path.join(ffcv_dir, ffcv_dest)
+    # always create if ffcv_write_dataset is true
+    if ffcv_write_dataset:
+        if dist.get_local_rank() == 0:
+            if datadir is None:
+                raise ValueError('datadir is required if use_synthetic is False and ffcv_write_dataset is True.')
+            ds = datasets.CIFAR10(
+                datadir,
+                train=is_train,
+                download=download,
+            )
+
+            write_ffcv_dataset(dataset=ds, write_path=dataset_filepath)
+
+        # Wait for the local rank 0 to be done creating the dataset in ffcv format.
+        dist.barrier()
+
+    if not os.path.exists(dataset_filepath):
+        raise ValueError(
+            f'Dataset file containing samples not found at {dataset_filepath}. Use ffcv_dir flag to point to a dir containing {dataset_filepath}.'
+        )
+
+    # Please note that this mean/std is different from the mean/std used for regular PyTorch dataloader as
+    # ToTensor does the normalization for PyTorch dataloaders.
+    cifar10_mean_ffcv = [125.307, 122.961, 113.8575]
```

Comment:
superfluous brackets (?)

Reply:
I'm going to assume it's supposed to be a list and not a tuple and keep those